askup/tests/test_functional/test_general.py

```diff
@@ -1443,3 +1459,169 @@ def test_user_has_no_questions(self):
+
+        response = self.get_my_subscriptions(self)
+        self.assertNotContains(response, 'UNSUBSCRIBE')
+
+    @client_user('student01', 'student01')
+    def test_subscription_success(self):
+        """
+        Test successful subscription.
+        """
+        qset_id = 4  # Qset 1-1 from the mockups
+        # Subscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'subscribe').json()
+        self.assertEqual(response['result'], 'success')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'unsubscribe'}
+            )
+        )
+
+        # Unsubscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'unsubscribe').json()
+        self.assertEqual(response['result'], 'success')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'subscribe'}
+            )
+        )
+
+    @client_user('student01', 'student01')
+    def test_subscription_fail(self):
+        """
+        Test failed on permission basis subscription.
+        """
+        qset_id = 10  # Qset 4-1 from the mockups
+        # Subscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'subscribe').json()
+        self.assertEqual(response['result'], 'fail')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'subscribe'}
+            )
+        )
+
+        # Unsubscribe current user to the selected qset
+        response = self.change_qset_subscription(self, qset_id, 'unsubscribe').json()
+        self.assertEqual(response['result'], 'fail')
+        self.assertEqual(
+            response['url'],
+            reverse(
+                'askup:qset_subscription',
+                kwargs={'qset_id': qset_id, 'subscribe': 'unsubscribe'}
+            )
+        )
+
+
+class TestSubscriptionsMailing(TestCase):
+    """
+    Testing the subject subscription mailing functionality.
+    """
+
+    fixtures = ['groups', 'mockup_data']
+
+    def setUp(self):
+        """
+        Set up the test conditions.
+        """
+        settings.DEBUG = False
+
+    @client_user('student01', 'student01')
+    @patch('askup.utils.general.do_send_subscriptions_to_recipients')
+    def test_subscriptions_sending_no_actual_subscriptions(self, mock_do_send_subscriptions_to_recipients):
+        """
+        Test subscriptions sending.
+        """
+        send_subscription_emails()
+        mock_do_send_subscriptions_to_recipients.return_value = None
```

Comment:
Is this necessary, given it wasn't called? Also, I believe it would have no effect if set after the actual method to be tested.

Reply:
Yeah. This two lines are actually working (1596, 1597), but I agree with you that the readability of the method could be by moving those into the beginning.
Thank you :+1: